### DISCUSSION TITLE
Deprecated payloader and simple solution.

### DISCUSSION THREAD

**Discussion by janiversen** [pymodbus-dev/pymodbus#2525] [General]
The binary payloder have been marked as deprecated since v3.8.0 and will be removed in v3.9.0

There have been quite a number of questions on how to replace it, so I thought it better to make a formal answer:

with v3.8.x binarypayloader is still part of the library, however simple usage like e.g reading/writing int32,int64,float32,float64 and strings are easily converted with the new convert_to/from_registers and those should be used.

With v3.9.0 and beyond binarypayloader is no longer available, for apps the prefer to read/write a chunk of registers, convert_combined_to/from_registers will be available. These function accept a list of tuples (datatype, count) which are used to convert correctly.

With v3.9.0 apps that uses datatypes not matched by registers like e.g. int8 or bits less than 16 will need to do their own conversions.

The new functions will be available on dev/wait_next_api in good time before release of v3.9.0 to allow for testing or early conversion. Progress can be followed on open pull requests and/or dev.

**Comment by saardrimer** [comment]
Thanks for the extra information and a very useful library! :)

I'm updating my script to use 3.8 and the deprecation messages are all over the place. 

```
BinaryPayloadBuilder is deprecated and will be removed in v3.9.0 !
Please use "client.convert_from_registers()" or "client.convert_to_registers"
See documentation: "https://pymodbus.readthedocs.io/en/latest/source/client.html#pymodbus.client.mixin.ModbusClientMixin.convert_from_registers"
```

Since it's not obvious for me how to fix this (use  convert_[to_from]_registers?) in order to make my console clean and prepare for 3.9, I tried to consult the example I used in the past:

https://github.com/pymodbus-dev/pymodbus/blob/dev/examples/client_payload.py

but it still uses  `BinaryPayloadBuilder` :(

I'm certain that that example will be updated by the release of 3.9, but in the meantime is there a way to suppress these messages? (I realise that this might not be a library thing, but my other suppression attempts did not work and you'd know the best solution.)

Many thanks.

**Reply by janiversen** [reply]
I do not think there are a way apart from
- not updating
- stop using binarypayloader
- modify the code

**Reply by janiversen** [reply]
And of course the most elegant way is to loop over registers and call to/from multiple times one for each datatype.

Apart from a couple of very old devices, I have not seen an example where binarypayload did more than just allow 1 read request instead of multiple.

**Reply by saardrimer** [reply]
Thanks for pushing me to figure this out. For the sake of those, like me, that aren't that well versed in pymodbus/modbus, here's what I did.

I need payloads for this:

```python
context = ModbusSlaveContext(
    di=ModbusSequentialDataBlock(1, [0]),
    co=CoilsDataBlock(1, coil_payload),
    hr=ModbusSequentialDataBlock(1, holding_payload),
    ir=ModbusSequentialDataBlock(1, input_payload),
)
```

Where I previously generated them like this:

```python
input_payload = BinaryPayloadBuilder(byteorder=Endian.BIG, wordorder=Endian.BIG)
input_payload.add_16bit_uint(0x0001)
input_payload.add_string("Hello!!!")
input_payload.add_32bit_uint(0x000AA000)
```

and now replaced it with (it's just a list! :) ):

```python
input_payload = []
input_payload.extend(
    ModbusTcpClient.convert_to_registers(
        value=0x0001,
        data_type=ModbusTcpClient.DATATYPE.UINT16,
        word_order="big",
    )
)
input_payload.extend(
    ModbusTcpClient.convert_to_registers(
        value="Hello!!!",
        data_type=ModbusTcpClient.DATATYPE.STRING,
        word_order="big",
    )
)
input_payload.extend(
    ModbusTcpClient.convert_to_registers(
        value=0x000AA000,
        data_type=ModbusTcpClient.DATATYPE.UINT32,
        word_order="big",
    )
)
```

Elsewhere for `PayloadBinaryDecoder` I had:

```python
decoder = BinaryPayloadDecoder.fromRegisters(
    regs,
    byteorder=Endian.BIG,
    wordorder=Endian.BIG,
)
result = decoder.decode_32bit_uint()
```

And now:

```python
result = ModbusTcpClient.convert_from_registers(
    registers=regs,
    data_type=ModbusTcpClient.DATATYPE.UINT32,
    word_order="big"
)
```

**Comment by janiversen** [comment]
You are using the binary payload for the server, this is a very old fashioned way (but it still works)... you could use the simulator datastore a lot more efficiently.

word_order is WRONG !!! it does not take a list, but a single literal ("big" or "little"), I really wonder why your code works. Seems you try to specify byte_order as well.

**Reply by saardrimer** [reply]
> You are using the binary payload for the server, this is a very old fashioned way (but it still works)... you could use the simulator datastore a lot more efficiently.

Thanks; I'll look into it.

> word_order is WRONG !!! it does not take a list, but a single literal ("big" or "little"), I really wonder why your code works. Seems you try to specify byte_order as well.

I'll fix it. I suppose that it works because the functions

https://github.com/pymodbus-dev/pymodbus/blob/f3762bbe8397b4a92abe36c371c9d8a3217458da/pymodbus/client/mixin.py#L738

https://github.com/pymodbus-dev/pymodbus/blob/f3762bbe8397b4a92abe36c371c9d8a3217458da/pymodbus/client/mixin.py#L697

don't check that input, but wouldn't do what I want if I was after "little".

**Reply by janiversen** [reply]
struct.pack / struct.unpack should do that check, and does it at least in some versions.

**Comment by wills106** [comment]
> With v3.9.0 and beyond binarypayloader is no longer available, for apps the prefer to read/write a chunk of registers, convert_combined_to/from_registers will be available. These function accept a list of tuples (datatype, count) which are used to convert correctly.

Just wondering when this will start to appear in one of the branches? As I can't seem to find it yet?

**Reply by janiversen** [reply]
Seems you did not follow the rest of the thread. Combined is abandoned, since it can be easily replaced with multiple  convert calls.
